ChronoStore/src/storywriter.cpp

```diff
@@ -0,0 +1,497 @@
+#include <iostream>
+
+#define H5_SIZEOF_SSIZE_T H5_SIZEOF_LONG_LONG
+
+#include <hdf5.h>
+#include <fstream>
+#include <string>
+#include <cstring>
+#include <vector>
+#include <event.h>
+#include <chunkattr.h>
+#include <storywriter.h>
+
+#define DATASET_RANK 1 // Dataset dimension
+#define ATTRIBUTE_CHUNKMETADATA "ChunkMetadata"
+#define ATTRIBUTE_DATASET_MIN "Min"
+#define ATTRIBUTE_DATASET_MAX "Max"
+#define ATTRIBUTE_TOTAL_DATASET_STORY_EVENTS "TotalStoryEvents"
+#define ATTRIBUTE_CHUNKMETADATA_RANK 1
+#define DEBUG 0 // Set to 1 ito print H5 error messages to console
+#define CHUNK_SIZE 40000 // Number of events equal to page size of 4MB
+
+// Constructor storywriter
+storywriter::storywriter(){}
+
+// Destructor storywriter
+storywriter::~storywriter(){}
+
+/**
+ * Write/Append data to storyDataset. 
+ * @param storyChunk: Data to write to story dataset.
+ * @param STORY: story name
+ * @param CHRONICLE: chronicle name
+ * @return: 0 if successful, else -1, if failed.
+ */
+int storywriter::writeStoryChunk(std::vector<Event> storyChunk, const char* STORY, const char* CHRONICLE) {
+
+    // Disable automatic printing of HDF5 error stack to console
+    if(!DEBUG){
+        H5Eset_auto(H5E_DEFAULT, NULL, NULL);
+    }
+
+    hid_t   ChronicleH5, storyDataset, storyDatasetSpace, chunkPropId, hyperslabMemorySpace, eventType, chunkmetaAttribute, attr, attr1, attr_DatasetMin, attr_DatasetMax, attr_DatasetTotalStoryEvents;
+    hid_t   attributeId, attributeType, attributeDataSpace;
+    herr_t  status, ret;
+
+    Event* eventBuffer;
+    ChunkAttr* attributeDataBuffer;
+    ChunkAttr* extendChunkMetadataBuffer;
+
+    /* 
+    * Chunk Dimension and number of events in a story chunk
+    */
+    const hsize_t chunkDims[1] = {CHUNK_SIZE};
+    const hsize_t numberOfEvents = storyChunk.size() - 1;
+
+    /*
+    * Allocate memory for events buffer
+    * TODO: Modification required in case of variable sized events
+    */
+    eventBuffer = (Event*) malloc(numberOfEvents * sizeof(Event));
+    if(eventBuffer == NULL){
+        return -1;
+    }
+
+    /*
+    * Allocate memory for attribute buffer
+    */
+    attributeDataBuffer = (ChunkAttr*) malloc(((numberOfEvents / CHUNK_SIZE) + 1) * sizeof(ChunkAttr));
+    if(attributeDataBuffer == NULL){
+        if(eventBuffer != NULL) free(eventBuffer);
+        return -1;
+    }
+
+    /*
+    * Read data from storyChunk into eventBuffer
+    */
+    for(auto event = 0 ; event < numberOfEvents ; event++) {
+    eventBuffer[event].timeStamp = storyChunk[event].timeStamp;
+    strcpy(eventBuffer[event].data, storyChunk[event].data);
+    }
```

Comment:
One vector of events should be enough - either eventBuffer or storyChunk. You can modify the API to writestoryChunk(eventBuffer,...)

Reply:
Done